File: generator/ec2_instance_types.py
import regions
import util
import boto3
import threading
import mypy_boto3_ec2
import mypy_boto3_ec2.type_defs as ec2_types
from typing import List, Dict, Literal, Set

ALL_INSTANCE_TYPES: Set[str] = set()
ALL_INSTANCE_FAMILIES: Set[str] = set()
INSTANCE_TYPE_OFFERINGS: Dict[
    Literal["regions", "instances", "families"], Dict[str, Set[str] | List[str]]
] = {
    "regions": {},
    "instances": {},
}
FAMILY_OFFERINGS: Dict[str, Dict[Literal["sizes", "regions"], Set[str]]] = {}


def GetInstanceTypes(ec2: mypy_boto3_ec2.Client, *, next_token=""):
    response = (
        ec2.describe_instance_types(NextToken=next_token)
        if next_token
        else ec2.describe_instance_types()
    )
    for instance_type in response["InstanceTypes"]:
        util.WriteFile(
            f"../frontend/src/content/instance-types/{instance_type['InstanceType']}.json",
            instance_type,
        )
    if "NextToken" in response:
        GetInstanceTypes(ec2, next_token=response["NextToken"])
        return


def GetInstanceOfferings(session: boto3.Session):
    region = session.region_name
    print(f"Getting Instance Offerings for {region}")
    ec2 = session.client("ec2")
    regional_instance_types = []
    for instance in ec2.describe_instance_type_offerings()["InstanceTypeOfferings"]:
        instance_type = instance["InstanceType"]
        ALL_INSTANCE_TYPES.add(instance_type)
        family = instance_type.split(".")[0]
        size = instance_type.split(".")[1]
        ALL_INSTANCE_FAMILIES.add(family)
        if not family in FAMILY_OFFERINGS:
            FAMILY_OFFERINGS[family] = {
                "regions": set(),
                "sizes": set(),
            }
        FAMILY_OFFERINGS[family]["regions"].add(region)
        FAMILY_OFFERINGS[family]["sizes"].add(size)
        regional_instance_types.append(instance_type)
        if not instance_type in INSTANCE_TYPE_OFFERINGS["instances"].keys():
            INSTANCE_TYPE_OFFERINGS["instances"][instance_type] = set()
        INSTANCE_TYPE_OFFERINGS["instances"][instance_type].add(region)
    INSTANCE_TYPE_OFFERINGS["regions"][region] = regional_instance_types
    print(f"Finished getting instance offerings for {region}")


def GetInstanceTypesForAllRegions():
    sessions: Dict[str, boto3.Session] = regions.GetSessions()
    threads: List[threading.Thread] = list()
    for region in sessions:
        thread = threading.Thread(target=GetInstanceOfferings, args=(sessions[region],))
        threads.append(thread)
        thread.start()
    GetInstanceTypes(sessions["us-east-1"].client("ec2"))
    for thread in threads:
        thread.join()

    util.WriteFile("../data/all_instance_types.json", ALL_INSTANCE_TYPES)
    util.WriteFile("../data/all_instance_families.json", ALL_INSTANCE_FAMILIES)
    util.WriteFile("../data/instance_type_offerings.json", INSTANCE_TYPE_OFFERINGS)
    util.WriteTypedFile(
        "../data/family_offerings.ts",
        FAMILY_OFFERINGS,
        """interface offering {
  regions: string[],
  sizes: string[],
}

export const offerings: { [Key: string]: offering }""",
    )


if __name__ == "__main__":
    GetInstanceOfferings()

File: generator/lib/ec2-instance-types.ts
import {
  EC2,
  type InstanceTypeInfo,
  paginateDescribeInstanceTypes,
} from '@aws-sdk/client-ec2';
import { getAllRegions } from './regions';
import { writeContentDir, writeDataToFile } from './util';

export interface InstanceFamily {
  name: string;
  regions: string[];
  sizes: string[];
  currentGeneration: boolean;
  cpu: {
    architectures: string[];
    minThreads: number;
    maxCores: number;
  };
  memory: {
    minMib: number;
    maxMib: number;
  };
  // network: {

  // }
}

export async function getInstanceTypes() {
  let instanceTypes: { [key: string]: InstanceTypeInfo } = {};

  const regions = await getAllRegions();

  // create a map of all instance types by their region
  let instanceTypesByRegion: { [Key: string]: string[] } = {};
  let regionsByInstanceType: { [Key: string]: string[] } = {};
  let instanceFreeTierRegions: { [Key: string]: string[] } = {};

  const getInstanceTypesByRegion = async (region: string) => {
    const ec2 = new EC2({ region });
    console.log(`Getting instance types for region ${region}...`);
    // get all instance types in this region
    for await (const page of paginateDescribeInstanceTypes(
      { client: ec2 },
      {}
    )) {
      for (let instanceType of page.InstanceTypes!) {
        if (instanceTypesByRegion[region]) {
          instanceTypesByRegion[region].push(instanceType.InstanceType!);
        } else {
          instanceTypesByRegion[region] = [instanceType.InstanceType!];
        }
        if (regionsByInstanceType[instanceType.InstanceType!]) {
          regionsByInstanceType[instanceType.InstanceType!].push(region);
        } else {
          regionsByInstanceType[instanceType.InstanceType!] = [region];
        }
        if (instanceType.FreeTierEligible) {
          if (instanceFreeTierRegions[region]) {
            instanceFreeTierRegions[region].push(instanceType.InstanceType!);
          } else {
            instanceFreeTierRegions[region] = [instanceType.InstanceType!];
          }
        }
        if (region == 'us-east-1') {
          instanceTypes[instanceType.InstanceType!] = instanceType;
        }
      }
    }
    console.log(`Finished getting instance types for region ${region}...`);
  };
  let promises = [];
  for (let region of regions) {
    promises.push(getInstanceTypesByRegion(region));
  }
  for (let promise of promises) {
    await promise;
  }

  // create a map of all instance types by their instance family
  let instanceFamilies: { [Key: string]: InstanceFamily } = {};
  for (let instanceType of Object.values(instanceTypes)) {
    const family = instanceType.InstanceType!.split('.')[0];
    const size = instanceType.InstanceType!.split('.')[1];
    if (instanceFamilies[family] == undefined) {
      instanceFamilies[family] = {
        name: family,
        regions: regionsByInstanceType[instanceType.InstanceType!].sort(),
        sizes: [],
        currentGeneration: instanceType.CurrentGeneration!,
        cpu: {
          architectures: instanceType.ProcessorInfo?.SupportedArchitectures!,
          minThreads: instanceType.VCpuInfo?.DefaultVCpus!,
          maxCores: instanceType.VCpuInfo?.DefaultVCpus!,
        },
        memory: {
          minMib: instanceType.MemoryInfo?.SizeInMiB!,
          maxMib: instanceType.MemoryInfo?.SizeInMiB!,
        },
      };
    }
    instanceFamilies[family].sizes.push(size);
    instanceFamilies[family].sizes = instanceFamilies[family].sizes.sort();
    instanceFamilies[family].cpu.minThreads = Math.min(
      instanceType.VCpuInfo?.DefaultVCpus!,
      instanceFamilies[family].cpu.minThreads
    );
    instanceFamilies[family].cpu.maxCores = Math.max(
      instanceType.VCpuInfo?.DefaultVCpus!,
      instanceFamilies[family].cpu.maxCores
    );
    instanceFamilies[family].memory.minMib = Math.min(
      instanceType.MemoryInfo?.SizeInMiB!,
      instanceFamilies[family].memory.minMib
    );
    instanceFamilies[family].memory.maxMib = Math.max(
      instanceType.MemoryInfo?.SizeInMiB!,
      instanceFamilies[family].memory.maxMib
    );
  }

  writeDataToFile(
    'src/lib/instance-free-tier-regions.json',
    instanceFreeTierRegions
  );

  writeContentDir('instance-families', instanceFamilies);
  writeContentDir('instance-types', instanceTypes);
}
